Finish him.

cmeiklejohn opened the pull request:
Resolve the remaining dialyzer errors on repl.  Flawless victory.

cmeiklejohn committed:
Remove errors from calls which no longer match spec.

cmeiklejohn committed:
Refactor functions.

Refactor functions so Dialyzer can more easily determine the types
originating from particular callers.

Resolves:

riak_repl_keylist_client.erl:331: The pattern 'cancel_fullsync' can never match the type 'pause_fullsync'

cmeiklejohn committed:
Ignore unset_env calls and resulting warnings.

Ignore invalid specs to unset_env; they are valid with tuples and
execute correctly, however, moving forward we should not use it because
OTP has updated the spec to reflect only the specific calls they want to
support moving forward.

cmeiklejohn committed:
Provide return value.

Provide an actual return value to prevent the no local return error
message.

Resolves:

riak_repl_console.erl:288: The created fun has no local return

seancribbs commented:
At first glance these seem ok to me, but I'm not a repl domain expert. This leaves only these remaining errors:

```
riak_core_connection_mgr.erl:494: The pattern 'ok' can never match the type {'error',atom()}
riak_repl_aae_source.erl:574: The pattern <MsgType, Msg, State> can never match since previous clauses completely covered the type <7,binary(),#state{index::'undefined' | non_neg_integer(),indexns::'undefined' | [{_,_}],tree_pid::'undefined' | pid(),tree_mref::'undefined' | reference(),built::'undefined' | non_neg_integer(),timeout::'undefined' | pos_integer(),wire_ver::atom(),diff_batch_size::non_neg_integer(),local_lock::boolean(),owner::'undefined' | pid()}>
riak_repl_console.erl:284: The created fun has no local return
riak_repl_console.erl:285: The pattern {'ok', Members} can never match the type [{string(),non_neg_integer()}]
```

cmeiklejohn commented:
Those will be addressed in a separate PR.

Vagabond commented on src/riak_repl_keylist_client.erl:
```diff
@@ -257,17 +260,13 @@ send_keylist(continue, #state{kl_fh=FH0,transport=Transport,socket=Socket,kl_cou
-    end,
-    log_stop(Command, State),
+wait_ack(cancel_fullsync, #state{sitename=SiteName} = State) ->
+    NewState = application:unset_env(riak_repl, {progress, SiteName}),
```
This is not the correct value for NewState.

cmeiklejohn replied:
Nice catch.  This was resolved in 6bb6262.

Vagabond commented:
I don't understand the reason for this ignore:

```
riak_repl_keylist_client.erl:330: The pattern 'cancel_fullsync' can never match the type 'pause_fullsync'
```

Vagabond commented:
The rest of this looks fine.

cmeiklejohn commented:
As best as I could tell via Typer, the unset_env and set_env calls which are producing errors because Dialyzer feels they will never succeed, causes it to identify that the `cancel_fullsync` branch will never be reached.

Vagabond commented:
You could move the log message before the (un)set_env calls, that should shut things up without doing anything other than logging a few instructions earlier.

cmeiklejohn committed:
Resolve incorrect assignment of state.

cmeiklejohn committed:
Reorder log messages.

Reorder log messages to prevent the incorrectly specified set_env and
unset_env calls from generating false errors.

cmeiklejohn commented:
Addressed logging issues with e30d78f437555e9543bdca8af9d79cca06759150.

cmeiklejohn committed:
Remove incorrect specification.

Resolves:

riak_repl_console.erl:284: The created fun has no local return
riak_repl_console.erl:285: The pattern {'ok', Members} can never match the type [{string(),non_neg_integer()}]

cmeiklejohn committed:
Ignore warning.

Regardless of the correct type specifications on sync_connect and
sync_connect_status, dialyzer still believes that this function can
never return 'ok' because the ranch call is thought to have failed
because of the improper specification.

See existing ignores related to ranch_tcp.

cmeiklejohn committed:
Resolve incorrect type signature.

cmeiklejohn committed:
Fix incorrecly specified type signature.

cmeiklejohn committed:
Remove unused clause.

All callers of this function use encode_obj_msg, which never returns a
non-binary, so crash if that is the case.

cmeiklejohn committed:
Ignore cluster_info as it's a conditional dep.

cmeiklejohn commented:
Depends on https://github.com/basho/riak_kv/pull/882.

Vagabond commented:
+1 on this.
